### DISCUSSION TITLE
Prevent transient from exiting AND from acting upon the *transient* buffer

### DISCUSSION THREAD

**Discussion by precompute** [magit/transient#251] [Support] [answered]
I'm using Emacs 30 and Transient `0.4.1`.  I'm trying to make a Transient that can enable / disable modes.  

``` lisp
(transient-define-prefix test-transient ()
  "Transient test."
  ;; :transient-suffix     'transient--do-stay
  ;; :transient-non-suffix 'transient--do-warn
  ["Heading 1"
   [("o" olivetti-mode
     ;; :transient t
     :description (lambda () (if olivetti-mode
                                 (propertize "ON" 'face '(bold success))
                               (propertize "OFF" 'face '(bold error)))))]])
```

The Transient defined in this example exits after `o` is pressed.

Uncommenting `:transient t` or `transient--do-stay` prevents the Transient from exiting after the command is executed, but the command is executed in the *transient* buffer instead of the buffer (or window) the Transient was called from.  

How can I make the Transient stay after execution, and also act on the buffer (or window) it was called from?  Uncommenting `:transient t` or `transient--do-stay` works on commit `51c50d` (ancient, from December 2021) which I'm using on my old Emacs config.

(I use "(or window)" because I've previously been able to change the buffer in a window and still have the transient work on it)

**Comment by tarsius** [comment]
> the command is executed in the transient buffer instead of the buffer (or window) the Transient was called from.

How did you make that determination? I don't think the code you provided would be able to demonstrate that issue. (It might be *affected* by such an issue, but it would be hard to tell what is going on.)

I don't have `olivetti-mode`, so I haven't tried with this.

Here is some other code that would be able to demonstrate that issue, but, for me at least, it instead demonstrates that things work as expected.

```elisp
(defvar-local test-var t)

(transient-define-prefix test-transient ()
  [("s" test-set-test-var :transient t :description
    (lambda () (format "[stay] (from %s) %s" (current-buffer) test-var)))
   ("e" (lambda () (interactive) (test-set-test-var))
    :transient nil :description
    (lambda () (format "[exit] (from %s) %s" (current-buffer) test-var)))])

(defun test-set-test-var ()
  (interactive)
  (let ((v (random 100)))
    (message "(in %s) %s => %s" (current-buffer) test-var v)
    (setq test-var v)))
```

**Reply by precompute** [reply]
Sorry, I wrote a new message below instead of using the reply box here (just noticed the box).

**Reply by tarsius** [reply]
Not your fault. Github should work on making the UI more obvious. As it stands, more than 50% of what should have been "replies", are instead written as "new answers".

**Comment by precompute** [comment]
Calling `message` and enabling/disabling a mode are different, the mode acts on a buffer while the message is always in the echo area (buffer/window independent).

I've tried https://github.com/positron-solutions/transient-showcase as well, and every example there also only uses `message`.

You can try this instead:
``` lisp
(transient-define-prefix test-transient ()
  "Transient test."
  :transient-suffix     'transient--do-stay
  ;; :transient-non-suffix 'transient--do-warn
  ["Heading 1"
   [("o" olivetti-mode
     ;; :transient t
     :description (lambda () (if olivetti-mode
                                 (propertize "ON" 'face '(bold success))
                               (propertize "OFF" 'face '(bold error)))))
    ("l" display-line-numbers-mode
     ;; :transient t
     :description "Toggle Line Numbers")
    ("v" visual-line-mode
     :description "Visual Lines")
    ("m" mixed-pitch-mode
     :description "Mixed Pitch")]]
  ["Major Modes"
   [("0" "Clean Mode" clean-mode)]
   [("1" "Fundamental Mode" fundamental-mode)
    ("2" "Emacs Lisp Mode" emacs-lisp-mode)]])
```

**Reply by tarsius** [reply]
> Calling `message` and enabling/disabling a mode are different, the mode acts on a buffer while the message is always in the echo area (buffer/window independent).

Yes of course, but that's not really relevant. `message` is just there so we can observe the effect. We could also just call `(setq-local some-var 'some-unique-value)` and then later manually inspect in what buffer the value of `same-var` is `some-unique-value`.

But let's use a mode. This is a minimal example, whose effect is easy to observe:

```elisp
(transient-define-prefix test-transient ()
  [("r" ruler-mode :transient t :description
    (lambda ()
      (format "Ruler mode is %s in %s"
              (if ruler-mode "on" "off") (current-buffer))))])
```

This toggles the mode in the buffer that was current before (and actually also *while*) the transient buffer is being shown. If that isn't the case for you, then start with `emacs -Q`, and if not even that fixes it, provide detailed steps to reproduce.

But begin with uninstalling Transient, exiting Emacs, restarting Emacs, reinstalling Transient -- just in case.

**Reply by precompute** [reply, marked as answer]
Thank you for the example and the explanation!  `(current-buffer)` was not `*transient*` for me.

I spent some time enabling and disabling things, and it turns out this wasn't an issue with Transient.  It's instead an issue with `shackle-mode` (https://depp.brause.cc/shackle/), which I was using to enforce popup positions.  I'm migrating from DOOM and wanted something easy, I guess I'll try to use `display-buffer-alist` instead.

**Reply by tarsius** [reply]
I've added a note about that package to https://github.com/magit/transient/wiki/FAQ. This package was the culprit before.
